# Practice D. Increase the number of hidden layers - Practice B but with 2 hidden layers, each of 10 nodes and ReLU
#####  Dataset
- [Download dataset here](https://cocl.us/concrete_data)

In [1]:
# Import packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import keras
from keras.models import Sequential
from keras.layers import Dense

from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

# Read in the data
concrete_data = pd.read_csv("/Users/pc/Desktop/IBM AI Engineer/Introduction to Deep Learning & Neural Networks with Keras/concrete_data.csv")

# Split data into predictors and target
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != "Strength"]]
target = concrete_data["Strength"]

2023-08-16 15:11:11.876835: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [2]:
# Normalization of the data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
target = (target - target.mean()) / target.std()

# the number of predictors
n_cols = predictors_norm.shape[1]

### 3 hidden layers, each of 10 nodes and ReLU activation function

In [3]:
# define the regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation = "relu", input_shape = (n_cols, )))
    model.add(Dense(10, activation = "relu", input_shape = (n_cols, )))
    model.add(Dense(10, activation = "relu", input_shape = (n_cols, )))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer = "adam", loss = "mean_squared_error")
    return model

# build the model
model = regression_model()

In [4]:
MSE_list = []
trial_times = 50
for t in range(trial_times):
    # train / test split
    X_train, X_test, y_train, y_test = train_test_split(predictors, 
                                                    target, 
                                                    test_size=0.3, 
                                                    random_state=42)
    # fit the model
    model.fit(X_train, y_train, epochs = 50, verbose = 2)
    
    # predict values
    predictions = model.predict(X_test)

    # Mean Squared Errors between predictions and the actual values: y_test
    predictions = predictions.reshape(predictions.size, )
    MSE_errors = mean_squared_error(y_test, predictions)
    MSE_list.append(MSE_errors)
    #print("The MSE errors at the {} epoch is {}.".format(e + 1, MSE_errors))


Epoch 1/50
23/23 - 1s - loss: 437.0296 - 747ms/epoch - 32ms/step
Epoch 2/50
23/23 - 0s - loss: 197.2645 - 21ms/epoch - 904us/step
Epoch 3/50
23/23 - 0s - loss: 88.8075 - 20ms/epoch - 875us/step
Epoch 4/50
23/23 - 0s - loss: 43.9731 - 21ms/epoch - 898us/step
Epoch 5/50
23/23 - 0s - loss: 26.5883 - 19ms/epoch - 841us/step
Epoch 6/50
23/23 - 0s - loss: 18.5361 - 18ms/epoch - 787us/step
Epoch 7/50
23/23 - 0s - loss: 13.8976 - 18ms/epoch - 778us/step
Epoch 8/50
23/23 - 0s - loss: 11.8259 - 25ms/epoch - 1ms/step
Epoch 9/50
23/23 - 0s - loss: 10.7460 - 47ms/epoch - 2ms/step
Epoch 10/50
23/23 - 0s - loss: 8.9201 - 18ms/epoch - 788us/step
Epoch 11/50
23/23 - 0s - loss: 7.8878 - 20ms/epoch - 860us/step
Epoch 12/50
23/23 - 0s - loss: 7.2680 - 18ms/epoch - 764us/step
Epoch 13/50
23/23 - 0s - loss: 6.6960 - 18ms/epoch - 767us/step
Epoch 14/50
23/23 - 0s - loss: 5.8382 - 19ms/epoch - 823us/step
Epoch 15/50
23/23 - 0s - loss: 5.3322 - 20ms/epoch - 851us/step
Epoch 16/50
23/23 - 0s - loss: 4.6882 - 19

Epoch 28/50
23/23 - 0s - loss: 0.5210 - 17ms/epoch - 748us/step
Epoch 29/50
23/23 - 0s - loss: 1.0800 - 17ms/epoch - 750us/step
Epoch 30/50
23/23 - 0s - loss: 0.5623 - 17ms/epoch - 745us/step
Epoch 31/50
23/23 - 0s - loss: 0.5189 - 17ms/epoch - 739us/step
Epoch 32/50
23/23 - 0s - loss: 0.7265 - 17ms/epoch - 755us/step
Epoch 33/50
23/23 - 0s - loss: 1.0498 - 17ms/epoch - 730us/step
Epoch 34/50
23/23 - 0s - loss: 0.4699 - 17ms/epoch - 744us/step
Epoch 35/50
23/23 - 0s - loss: 0.5294 - 17ms/epoch - 750us/step
Epoch 36/50
23/23 - 0s - loss: 0.5127 - 21ms/epoch - 899us/step
Epoch 37/50
23/23 - 0s - loss: 0.5017 - 18ms/epoch - 796us/step
Epoch 38/50
23/23 - 0s - loss: 0.5018 - 18ms/epoch - 764us/step
Epoch 39/50
23/23 - 0s - loss: 0.6679 - 18ms/epoch - 778us/step
Epoch 40/50
23/23 - 0s - loss: 0.5983 - 17ms/epoch - 756us/step
Epoch 41/50
23/23 - 0s - loss: 0.5372 - 17ms/epoch - 736us/step
Epoch 42/50
23/23 - 0s - loss: 0.7328 - 29ms/epoch - 1ms/step
Epoch 43/50
23/23 - 0s - loss: 0.6493 - 21

Epoch 4/50
23/23 - 0s - loss: 0.7736 - 17ms/epoch - 751us/step
Epoch 5/50
23/23 - 0s - loss: 0.5377 - 17ms/epoch - 760us/step
Epoch 6/50
23/23 - 0s - loss: 0.9049 - 17ms/epoch - 748us/step
Epoch 7/50
23/23 - 0s - loss: 0.7439 - 17ms/epoch - 744us/step
Epoch 8/50
23/23 - 0s - loss: 0.5487 - 19ms/epoch - 806us/step
Epoch 9/50
23/23 - 0s - loss: 0.5309 - 33ms/epoch - 1ms/step
Epoch 10/50
23/23 - 0s - loss: 0.4906 - 29ms/epoch - 1ms/step
Epoch 11/50
23/23 - 0s - loss: 0.4416 - 24ms/epoch - 1ms/step
Epoch 12/50
23/23 - 0s - loss: 0.5280 - 17ms/epoch - 743us/step
Epoch 13/50
23/23 - 0s - loss: 0.4595 - 17ms/epoch - 754us/step
Epoch 14/50
23/23 - 0s - loss: 0.4896 - 17ms/epoch - 754us/step
Epoch 15/50
23/23 - 0s - loss: 0.5246 - 17ms/epoch - 750us/step
Epoch 16/50
23/23 - 0s - loss: 0.4959 - 17ms/epoch - 755us/step
Epoch 17/50
23/23 - 0s - loss: 0.4353 - 17ms/epoch - 750us/step
Epoch 18/50
23/23 - 0s - loss: 0.4707 - 17ms/epoch - 758us/step
Epoch 19/50
23/23 - 0s - loss: 0.6002 - 17ms/epoch -

Epoch 32/50
23/23 - 0s - loss: 0.4655 - 19ms/epoch - 816us/step
Epoch 33/50
23/23 - 0s - loss: 0.4152 - 18ms/epoch - 792us/step
Epoch 34/50
23/23 - 0s - loss: 0.4846 - 20ms/epoch - 854us/step
Epoch 35/50
23/23 - 0s - loss: 0.3937 - 18ms/epoch - 767us/step
Epoch 36/50
23/23 - 0s - loss: 0.4648 - 18ms/epoch - 765us/step
Epoch 37/50
23/23 - 0s - loss: 0.5510 - 17ms/epoch - 756us/step
Epoch 38/50
23/23 - 0s - loss: 0.4060 - 18ms/epoch - 787us/step
Epoch 39/50
23/23 - 0s - loss: 0.5012 - 19ms/epoch - 818us/step
Epoch 40/50
23/23 - 0s - loss: 0.4392 - 17ms/epoch - 745us/step
Epoch 41/50
23/23 - 0s - loss: 0.7233 - 17ms/epoch - 749us/step
Epoch 42/50
23/23 - 0s - loss: 0.5266 - 17ms/epoch - 756us/step
Epoch 43/50
23/23 - 0s - loss: 0.5874 - 17ms/epoch - 746us/step
Epoch 44/50
23/23 - 0s - loss: 0.5052 - 17ms/epoch - 751us/step
Epoch 45/50
23/23 - 0s - loss: 0.6153 - 17ms/epoch - 753us/step
Epoch 46/50
23/23 - 0s - loss: 0.4407 - 17ms/epoch - 753us/step
Epoch 47/50
23/23 - 0s - loss: 0.4215 - 

23/23 - 0s - loss: 0.4440 - 18ms/epoch - 777us/step
Epoch 9/50
23/23 - 0s - loss: 0.4411 - 19ms/epoch - 805us/step
Epoch 10/50
23/23 - 0s - loss: 0.4898 - 18ms/epoch - 784us/step
Epoch 11/50
23/23 - 0s - loss: 0.4071 - 17ms/epoch - 752us/step
Epoch 12/50
23/23 - 0s - loss: 0.4302 - 17ms/epoch - 753us/step
Epoch 13/50
23/23 - 0s - loss: 0.4084 - 17ms/epoch - 757us/step
Epoch 14/50
23/23 - 0s - loss: 0.4070 - 17ms/epoch - 758us/step
Epoch 15/50
23/23 - 0s - loss: 0.4138 - 18ms/epoch - 768us/step
Epoch 16/50
23/23 - 0s - loss: 0.4021 - 17ms/epoch - 749us/step
Epoch 17/50
23/23 - 0s - loss: 0.4911 - 17ms/epoch - 757us/step
Epoch 18/50
23/23 - 0s - loss: 0.3667 - 17ms/epoch - 757us/step
Epoch 19/50
23/23 - 0s - loss: 0.4654 - 18ms/epoch - 799us/step
Epoch 20/50
23/23 - 0s - loss: 0.3736 - 18ms/epoch - 770us/step
Epoch 21/50
23/23 - 0s - loss: 0.4200 - 18ms/epoch - 780us/step
Epoch 22/50
23/23 - 0s - loss: 0.3597 - 18ms/epoch - 787us/step
Epoch 23/50
23/23 - 0s - loss: 0.3745 - 17ms/epoch - 

Epoch 35/50
23/23 - 0s - loss: 0.4374 - 18ms/epoch - 775us/step
Epoch 36/50
23/23 - 0s - loss: 0.3888 - 17ms/epoch - 759us/step
Epoch 37/50
23/23 - 0s - loss: 0.3585 - 18ms/epoch - 780us/step
Epoch 38/50
23/23 - 0s - loss: 0.3554 - 18ms/epoch - 781us/step
Epoch 39/50
23/23 - 0s - loss: 0.3897 - 19ms/epoch - 810us/step
Epoch 40/50
23/23 - 0s - loss: 0.4158 - 18ms/epoch - 774us/step
Epoch 41/50
23/23 - 0s - loss: 0.4298 - 19ms/epoch - 820us/step
Epoch 42/50
23/23 - 0s - loss: 0.4942 - 18ms/epoch - 776us/step
Epoch 43/50
23/23 - 0s - loss: 0.4084 - 24ms/epoch - 1ms/step
Epoch 44/50
23/23 - 0s - loss: 0.4387 - 19ms/epoch - 818us/step
Epoch 45/50
23/23 - 0s - loss: 0.5121 - 17ms/epoch - 751us/step
Epoch 46/50
23/23 - 0s - loss: 0.4174 - 17ms/epoch - 750us/step
Epoch 47/50
23/23 - 0s - loss: 0.3828 - 18ms/epoch - 792us/step
Epoch 48/50
23/23 - 0s - loss: 0.4570 - 17ms/epoch - 759us/step
Epoch 49/50
23/23 - 0s - loss: 0.3799 - 17ms/epoch - 756us/step
Epoch 50/50
23/23 - 0s - loss: 0.3609 - 21

23/23 - 0s - loss: 0.2261 - 18ms/epoch - 770us/step
Epoch 12/50
23/23 - 0s - loss: 0.2098 - 18ms/epoch - 762us/step
Epoch 13/50
23/23 - 0s - loss: 0.1955 - 18ms/epoch - 772us/step
Epoch 14/50
23/23 - 0s - loss: 0.2161 - 20ms/epoch - 850us/step
Epoch 15/50
23/23 - 0s - loss: 0.2367 - 18ms/epoch - 791us/step
Epoch 16/50
23/23 - 0s - loss: 0.1936 - 18ms/epoch - 801us/step
Epoch 17/50
23/23 - 0s - loss: 0.2256 - 23ms/epoch - 982us/step
Epoch 18/50
23/23 - 0s - loss: 0.2686 - 20ms/epoch - 885us/step
Epoch 19/50
23/23 - 0s - loss: 0.2390 - 17ms/epoch - 739us/step
Epoch 20/50
23/23 - 0s - loss: 0.2171 - 17ms/epoch - 744us/step
Epoch 21/50
23/23 - 0s - loss: 0.2185 - 17ms/epoch - 740us/step
Epoch 22/50
23/23 - 0s - loss: 0.2419 - 18ms/epoch - 766us/step
Epoch 23/50
23/23 - 0s - loss: 0.2476 - 18ms/epoch - 769us/step
Epoch 24/50
23/23 - 0s - loss: 0.2047 - 20ms/epoch - 849us/step
Epoch 25/50
23/23 - 0s - loss: 0.2125 - 19ms/epoch - 839us/step
Epoch 26/50
23/23 - 0s - loss: 0.2000 - 21ms/epoch -

Epoch 38/50
23/23 - 0s - loss: 0.3355 - 22ms/epoch - 944us/step
Epoch 39/50
23/23 - 0s - loss: 0.2158 - 40ms/epoch - 2ms/step
Epoch 40/50
23/23 - 0s - loss: 0.1791 - 27ms/epoch - 1ms/step
Epoch 41/50
23/23 - 0s - loss: 0.1842 - 19ms/epoch - 816us/step
Epoch 42/50
23/23 - 0s - loss: 0.1781 - 19ms/epoch - 829us/step
Epoch 43/50
23/23 - 0s - loss: 0.1856 - 19ms/epoch - 812us/step
Epoch 44/50
23/23 - 0s - loss: 0.1800 - 19ms/epoch - 822us/step
Epoch 45/50
23/23 - 0s - loss: 0.2055 - 19ms/epoch - 809us/step
Epoch 46/50
23/23 - 0s - loss: 0.1885 - 19ms/epoch - 814us/step
Epoch 47/50
23/23 - 0s - loss: 0.2168 - 19ms/epoch - 817us/step
Epoch 48/50
23/23 - 0s - loss: 0.2302 - 19ms/epoch - 813us/step
Epoch 49/50
23/23 - 0s - loss: 0.2394 - 19ms/epoch - 823us/step
Epoch 50/50
23/23 - 0s - loss: 0.2061 - 20ms/epoch - 857us/step
10/10 [==============================] - 0s 841us/step
Epoch 1/50
23/23 - 0s - loss: 0.1880 - 20ms/epoch - 850us/step
Epoch 2/50
23/23 - 0s - loss: 0.2833 - 18ms/epoch - 79

Epoch 15/50
23/23 - 0s - loss: 0.1798 - 22ms/epoch - 975us/step
Epoch 16/50
23/23 - 0s - loss: 0.2192 - 20ms/epoch - 852us/step
Epoch 17/50
23/23 - 0s - loss: 0.2067 - 23ms/epoch - 992us/step
Epoch 18/50
23/23 - 0s - loss: 0.2421 - 21ms/epoch - 895us/step
Epoch 19/50
23/23 - 0s - loss: 0.1849 - 20ms/epoch - 879us/step
Epoch 20/50
23/23 - 0s - loss: 0.1935 - 25ms/epoch - 1ms/step
Epoch 21/50
23/23 - 0s - loss: 0.1782 - 27ms/epoch - 1ms/step
Epoch 22/50
23/23 - 0s - loss: 0.2182 - 21ms/epoch - 932us/step
Epoch 23/50
23/23 - 0s - loss: 0.1805 - 22ms/epoch - 974us/step
Epoch 24/50
23/23 - 0s - loss: 0.2076 - 21ms/epoch - 908us/step
Epoch 25/50
23/23 - 0s - loss: 0.1657 - 21ms/epoch - 927us/step
Epoch 26/50
23/23 - 0s - loss: 0.1926 - 21ms/epoch - 922us/step
Epoch 27/50
23/23 - 0s - loss: 0.1883 - 21ms/epoch - 910us/step
Epoch 28/50
23/23 - 0s - loss: 0.1923 - 22ms/epoch - 974us/step
Epoch 29/50
23/23 - 0s - loss: 0.2257 - 64ms/epoch - 3ms/step
Epoch 30/50
23/23 - 0s - loss: 0.2006 - 28ms/e

Epoch 42/50
23/23 - 0s - loss: 0.2491 - 17ms/epoch - 755us/step
Epoch 43/50
23/23 - 0s - loss: 0.1813 - 17ms/epoch - 747us/step
Epoch 44/50
23/23 - 0s - loss: 0.1777 - 17ms/epoch - 758us/step
Epoch 45/50
23/23 - 0s - loss: 0.2027 - 17ms/epoch - 747us/step
Epoch 46/50
23/23 - 0s - loss: 0.2477 - 17ms/epoch - 745us/step
Epoch 47/50
23/23 - 0s - loss: 0.1959 - 18ms/epoch - 790us/step
Epoch 48/50
23/23 - 0s - loss: 0.1780 - 19ms/epoch - 835us/step
Epoch 49/50
23/23 - 0s - loss: 0.1809 - 19ms/epoch - 842us/step
Epoch 50/50
23/23 - 0s - loss: 0.1913 - 19ms/epoch - 822us/step
10/10 [==============================] - 0s 776us/step
Epoch 1/50
23/23 - 0s - loss: 0.1704 - 21ms/epoch - 894us/step
Epoch 2/50
23/23 - 0s - loss: 0.1793 - 18ms/epoch - 795us/step
Epoch 3/50
23/23 - 0s - loss: 0.1871 - 19ms/epoch - 818us/step
Epoch 4/50
23/23 - 0s - loss: 0.1989 - 18ms/epoch - 794us/step
Epoch 5/50
23/23 - 0s - loss: 0.1962 - 19ms/epoch - 833us/step
Epoch 6/50
23/23 - 0s - loss: 0.1945 - 18ms/epoch - 80

Epoch 18/50
23/23 - 0s - loss: 0.1764 - 17ms/epoch - 758us/step
Epoch 19/50
23/23 - 0s - loss: 0.1659 - 18ms/epoch - 795us/step
Epoch 20/50
23/23 - 0s - loss: 0.1652 - 17ms/epoch - 748us/step
Epoch 21/50
23/23 - 0s - loss: 0.1873 - 18ms/epoch - 761us/step
Epoch 22/50
23/23 - 0s - loss: 0.1619 - 18ms/epoch - 803us/step
Epoch 23/50
23/23 - 0s - loss: 0.1576 - 21ms/epoch - 892us/step
Epoch 24/50
23/23 - 0s - loss: 0.1634 - 18ms/epoch - 778us/step
Epoch 25/50
23/23 - 0s - loss: 0.1791 - 17ms/epoch - 749us/step
Epoch 26/50
23/23 - 0s - loss: 0.1633 - 17ms/epoch - 742us/step
Epoch 27/50
23/23 - 0s - loss: 0.1779 - 18ms/epoch - 775us/step
Epoch 28/50
23/23 - 0s - loss: 0.1582 - 17ms/epoch - 758us/step
Epoch 29/50
23/23 - 0s - loss: 0.1618 - 17ms/epoch - 756us/step
Epoch 30/50
23/23 - 0s - loss: 0.1703 - 17ms/epoch - 750us/step
Epoch 31/50
23/23 - 0s - loss: 0.2111 - 18ms/epoch - 776us/step
Epoch 32/50
23/23 - 0s - loss: 0.1709 - 17ms/epoch - 754us/step
Epoch 33/50
23/23 - 0s - loss: 0.1876 - 

Epoch 45/50
23/23 - 0s - loss: 0.1980 - 19ms/epoch - 814us/step
Epoch 46/50
23/23 - 0s - loss: 0.1789 - 19ms/epoch - 835us/step
Epoch 47/50
23/23 - 0s - loss: 0.1780 - 19ms/epoch - 814us/step
Epoch 48/50
23/23 - 0s - loss: 0.1769 - 18ms/epoch - 797us/step
Epoch 49/50
23/23 - 0s - loss: 0.1772 - 18ms/epoch - 781us/step
Epoch 50/50
23/23 - 0s - loss: 0.1860 - 18ms/epoch - 766us/step
10/10 [==============================] - 0s 903us/step
Epoch 1/50
23/23 - 0s - loss: 0.1732 - 21ms/epoch - 896us/step
Epoch 2/50
23/23 - 0s - loss: 0.1907 - 19ms/epoch - 830us/step
Epoch 3/50
23/23 - 0s - loss: 0.1868 - 20ms/epoch - 852us/step
Epoch 4/50
23/23 - 0s - loss: 0.1625 - 21ms/epoch - 906us/step
Epoch 5/50
23/23 - 0s - loss: 0.1854 - 23ms/epoch - 1ms/step
Epoch 6/50
23/23 - 0s - loss: 0.1739 - 20ms/epoch - 879us/step
Epoch 7/50
23/23 - 0s - loss: 0.1630 - 19ms/epoch - 839us/step
Epoch 8/50
23/23 - 0s - loss: 0.1641 - 19ms/epoch - 832us/step
Epoch 9/50
23/23 - 0s - loss: 0.1687 - 21ms/epoch - 907us/s

Epoch 22/50
23/23 - 0s - loss: 0.1639 - 17ms/epoch - 741us/step
Epoch 23/50
23/23 - 0s - loss: 0.1665 - 17ms/epoch - 758us/step
Epoch 24/50
23/23 - 0s - loss: 0.1874 - 17ms/epoch - 747us/step
Epoch 25/50
23/23 - 0s - loss: 0.1707 - 17ms/epoch - 759us/step
Epoch 26/50
23/23 - 0s - loss: 0.1578 - 17ms/epoch - 752us/step
Epoch 27/50
23/23 - 0s - loss: 0.1582 - 17ms/epoch - 743us/step
Epoch 28/50
23/23 - 0s - loss: 0.1764 - 18ms/epoch - 796us/step
Epoch 29/50
23/23 - 0s - loss: 0.1484 - 18ms/epoch - 770us/step
Epoch 30/50
23/23 - 0s - loss: 0.1633 - 17ms/epoch - 742us/step
Epoch 31/50
23/23 - 0s - loss: 0.1620 - 17ms/epoch - 740us/step
Epoch 32/50
23/23 - 0s - loss: 0.1896 - 17ms/epoch - 739us/step
Epoch 33/50
23/23 - 0s - loss: 0.1719 - 17ms/epoch - 747us/step
Epoch 34/50
23/23 - 0s - loss: 0.1662 - 23ms/epoch - 1ms/step
Epoch 35/50
23/23 - 0s - loss: 0.1629 - 18ms/epoch - 764us/step
Epoch 36/50
23/23 - 0s - loss: 0.1607 - 24ms/epoch - 1ms/step
Epoch 37/50
23/23 - 0s - loss: 0.1525 - 17ms

Epoch 49/50
23/23 - 0s - loss: 0.1796 - 18ms/epoch - 800us/step
Epoch 50/50
23/23 - 0s - loss: 0.2120 - 18ms/epoch - 769us/step
10/10 [==============================] - 0s 1ms/step
Epoch 1/50
23/23 - 0s - loss: 0.1770 - 22ms/epoch - 978us/step
Epoch 2/50
23/23 - 0s - loss: 0.1576 - 20ms/epoch - 880us/step
Epoch 3/50
23/23 - 0s - loss: 0.1883 - 19ms/epoch - 811us/step
Epoch 4/50
23/23 - 0s - loss: 0.1503 - 17ms/epoch - 758us/step
Epoch 5/50
23/23 - 0s - loss: 0.1826 - 17ms/epoch - 755us/step
Epoch 6/50
23/23 - 0s - loss: 0.1705 - 17ms/epoch - 738us/step
Epoch 7/50
23/23 - 0s - loss: 0.1774 - 23ms/epoch - 983us/step
Epoch 8/50
23/23 - 0s - loss: 0.1967 - 18ms/epoch - 790us/step
Epoch 9/50
23/23 - 0s - loss: 0.1629 - 17ms/epoch - 741us/step
Epoch 10/50
23/23 - 0s - loss: 0.1541 - 18ms/epoch - 796us/step
Epoch 11/50
23/23 - 0s - loss: 0.1458 - 17ms/epoch - 728us/step
Epoch 12/50
23/23 - 0s - loss: 0.1461 - 17ms/epoch - 748us/step
Epoch 13/50
23/23 - 0s - loss: 0.1847 - 17ms/epoch - 757us/s

Epoch 26/50
23/23 - 0s - loss: 0.1535 - 19ms/epoch - 818us/step
Epoch 27/50
23/23 - 0s - loss: 0.1540 - 19ms/epoch - 829us/step
Epoch 28/50
23/23 - 0s - loss: 0.1775 - 19ms/epoch - 808us/step
Epoch 29/50
23/23 - 0s - loss: 0.1829 - 19ms/epoch - 812us/step
Epoch 30/50
23/23 - 0s - loss: 0.1525 - 19ms/epoch - 805us/step
Epoch 31/50
23/23 - 0s - loss: 0.1889 - 19ms/epoch - 818us/step
Epoch 32/50
23/23 - 0s - loss: 0.1532 - 19ms/epoch - 812us/step
Epoch 33/50
23/23 - 0s - loss: 0.1476 - 19ms/epoch - 807us/step
Epoch 34/50
23/23 - 0s - loss: 0.1492 - 19ms/epoch - 811us/step
Epoch 35/50
23/23 - 0s - loss: 0.1547 - 19ms/epoch - 825us/step
Epoch 36/50
23/23 - 0s - loss: 0.1462 - 19ms/epoch - 820us/step
Epoch 37/50
23/23 - 0s - loss: 0.1497 - 18ms/epoch - 803us/step
Epoch 38/50
23/23 - 0s - loss: 0.1470 - 19ms/epoch - 822us/step
Epoch 39/50
23/23 - 0s - loss: 0.1551 - 19ms/epoch - 804us/step
Epoch 40/50
23/23 - 0s - loss: 0.1446 - 19ms/epoch - 834us/step
Epoch 41/50
23/23 - 0s - loss: 0.1522 - 

Epoch 2/50
23/23 - 0s - loss: 0.1464 - 17ms/epoch - 760us/step
Epoch 3/50
23/23 - 0s - loss: 0.1529 - 18ms/epoch - 778us/step
Epoch 4/50
23/23 - 0s - loss: 0.1497 - 18ms/epoch - 761us/step
Epoch 5/50
23/23 - 0s - loss: 0.1488 - 17ms/epoch - 751us/step
Epoch 6/50
23/23 - 0s - loss: 0.1514 - 20ms/epoch - 876us/step
Epoch 7/50
23/23 - 0s - loss: 0.1500 - 23ms/epoch - 987us/step
Epoch 8/50
23/23 - 0s - loss: 0.1400 - 18ms/epoch - 789us/step
Epoch 9/50
23/23 - 0s - loss: 0.1623 - 18ms/epoch - 791us/step
Epoch 10/50
23/23 - 0s - loss: 0.1936 - 20ms/epoch - 849us/step
Epoch 11/50
23/23 - 0s - loss: 0.1481 - 18ms/epoch - 777us/step
Epoch 12/50
23/23 - 0s - loss: 0.1427 - 18ms/epoch - 778us/step
Epoch 13/50
23/23 - 0s - loss: 0.1729 - 18ms/epoch - 769us/step
Epoch 14/50
23/23 - 0s - loss: 0.1457 - 18ms/epoch - 767us/step
Epoch 15/50
23/23 - 0s - loss: 0.1521 - 17ms/epoch - 753us/step
Epoch 16/50
23/23 - 0s - loss: 0.1820 - 17ms/epoch - 759us/step
Epoch 17/50
23/23 - 0s - loss: 0.1885 - 17ms/epo

Epoch 29/50
23/23 - 0s - loss: 0.1488 - 19ms/epoch - 832us/step
Epoch 30/50
23/23 - 0s - loss: 0.1364 - 20ms/epoch - 855us/step
Epoch 31/50
23/23 - 0s - loss: 0.1344 - 19ms/epoch - 839us/step
Epoch 32/50
23/23 - 0s - loss: 0.1266 - 20ms/epoch - 860us/step
Epoch 33/50
23/23 - 0s - loss: 0.1273 - 18ms/epoch - 785us/step
Epoch 34/50
23/23 - 0s - loss: 0.1626 - 19ms/epoch - 813us/step
Epoch 35/50
23/23 - 0s - loss: 0.1645 - 18ms/epoch - 800us/step
Epoch 36/50
23/23 - 0s - loss: 0.1321 - 19ms/epoch - 813us/step
Epoch 37/50
23/23 - 0s - loss: 0.1480 - 18ms/epoch - 802us/step
Epoch 38/50
23/23 - 0s - loss: 0.1656 - 19ms/epoch - 823us/step
Epoch 39/50
23/23 - 0s - loss: 0.1643 - 18ms/epoch - 784us/step
Epoch 40/50
23/23 - 0s - loss: 0.1632 - 27ms/epoch - 1ms/step
Epoch 41/50
23/23 - 0s - loss: 0.1473 - 29ms/epoch - 1ms/step
Epoch 42/50
23/23 - 0s - loss: 0.1367 - 20ms/epoch - 884us/step
Epoch 43/50
23/23 - 0s - loss: 0.1415 - 19ms/epoch - 811us/step
Epoch 44/50
23/23 - 0s - loss: 0.1669 - 21ms

Epoch 5/50
23/23 - 0s - loss: 0.1454 - 22ms/epoch - 947us/step
Epoch 6/50
23/23 - 0s - loss: 0.1305 - 19ms/epoch - 842us/step
Epoch 7/50
23/23 - 0s - loss: 0.1284 - 19ms/epoch - 829us/step
Epoch 8/50
23/23 - 0s - loss: 0.1361 - 19ms/epoch - 818us/step
Epoch 9/50
23/23 - 0s - loss: 0.1347 - 19ms/epoch - 818us/step
Epoch 10/50
23/23 - 0s - loss: 0.1366 - 19ms/epoch - 806us/step
Epoch 11/50
23/23 - 0s - loss: 0.1387 - 19ms/epoch - 830us/step
Epoch 12/50
23/23 - 0s - loss: 0.1267 - 18ms/epoch - 801us/step
Epoch 13/50
23/23 - 0s - loss: 0.1273 - 20ms/epoch - 853us/step
Epoch 14/50
23/23 - 0s - loss: 0.1225 - 19ms/epoch - 828us/step
Epoch 15/50
23/23 - 0s - loss: 0.1289 - 19ms/epoch - 815us/step
Epoch 16/50
23/23 - 0s - loss: 0.1288 - 19ms/epoch - 820us/step
Epoch 17/50
23/23 - 0s - loss: 0.1420 - 19ms/epoch - 820us/step
Epoch 18/50
23/23 - 0s - loss: 0.1393 - 17ms/epoch - 753us/step
Epoch 19/50
23/23 - 0s - loss: 0.1283 - 20ms/epoch - 879us/step
Epoch 20/50
23/23 - 0s - loss: 0.1325 - 20ms/

Epoch 32/50
23/23 - 0s - loss: 0.1382 - 17ms/epoch - 750us/step
Epoch 33/50
23/23 - 0s - loss: 0.1331 - 17ms/epoch - 753us/step
Epoch 34/50
23/23 - 0s - loss: 0.1577 - 18ms/epoch - 776us/step
Epoch 35/50
23/23 - 0s - loss: 0.1500 - 18ms/epoch - 767us/step
Epoch 36/50
23/23 - 0s - loss: 0.1278 - 18ms/epoch - 784us/step
Epoch 37/50
23/23 - 0s - loss: 0.1274 - 17ms/epoch - 755us/step
Epoch 38/50
23/23 - 0s - loss: 0.1371 - 26ms/epoch - 1ms/step
Epoch 39/50
23/23 - 0s - loss: 0.1357 - 24ms/epoch - 1ms/step
Epoch 40/50
23/23 - 0s - loss: 0.1243 - 19ms/epoch - 817us/step
Epoch 41/50
23/23 - 0s - loss: 0.1368 - 19ms/epoch - 829us/step
Epoch 42/50
23/23 - 0s - loss: 0.1401 - 18ms/epoch - 796us/step
Epoch 43/50
23/23 - 0s - loss: 0.1264 - 18ms/epoch - 798us/step
Epoch 44/50
23/23 - 0s - loss: 0.1259 - 18ms/epoch - 798us/step
Epoch 45/50
23/23 - 0s - loss: 0.1340 - 19ms/epoch - 831us/step
Epoch 46/50
23/23 - 0s - loss: 0.1365 - 18ms/epoch - 803us/step
Epoch 47/50
23/23 - 0s - loss: 0.1716 - 18ms

Epoch 8/50
23/23 - 0s - loss: 0.1485 - 20ms/epoch - 867us/step
Epoch 9/50
23/23 - 0s - loss: 0.1278 - 21ms/epoch - 906us/step
Epoch 10/50
23/23 - 0s - loss: 0.1236 - 20ms/epoch - 853us/step
Epoch 11/50
23/23 - 0s - loss: 0.1219 - 20ms/epoch - 856us/step
Epoch 12/50
23/23 - 0s - loss: 0.1261 - 20ms/epoch - 890us/step
Epoch 13/50
23/23 - 0s - loss: 0.1375 - 22ms/epoch - 970us/step
Epoch 14/50
23/23 - 0s - loss: 0.1321 - 20ms/epoch - 886us/step
Epoch 15/50
23/23 - 0s - loss: 0.1260 - 20ms/epoch - 885us/step
Epoch 16/50
23/23 - 0s - loss: 0.1233 - 21ms/epoch - 896us/step
Epoch 17/50
23/23 - 0s - loss: 0.1299 - 21ms/epoch - 925us/step
Epoch 18/50
23/23 - 0s - loss: 0.1239 - 20ms/epoch - 872us/step
Epoch 19/50
23/23 - 0s - loss: 0.1333 - 21ms/epoch - 894us/step
Epoch 20/50
23/23 - 0s - loss: 0.1195 - 21ms/epoch - 901us/step
Epoch 21/50
23/23 - 0s - loss: 0.1320 - 21ms/epoch - 892us/step
Epoch 22/50
23/23 - 0s - loss: 0.1176 - 21ms/epoch - 927us/step
Epoch 23/50
23/23 - 0s - loss: 0.1195 - 21

### Comparison to the mean of the MSEs compred to which from A

In [5]:
print("The final MSE at the 50th time is {}.".format(MSE_list[-1]))
print("The mean of the MSE is {}.".format(np.mean(MSE_list)))
print("The standard deviation of the MSE is {}".format(np.std(MSE_list)))


The final MSE at the 50th time is 0.1258030593557269.
The mean of the MSE is 0.2975087229090375.
The standard deviation of the MSE is 0.23740818880204367


## How does the mean of the MSEs compare to which from Step B?

#### Mean of MSEs from Step D: 0.5262634357577329
#### Mean of MSEs from Step B: 0.2975087229090375
### Thus, the mean of the MSEs is decreased compare to which from Step B, actually also the mean is also less than which from Step C.